aten/src/ATen/core/ivalue_inl.h

```diff
@@ -346,18 +348,29 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   /**
    * Wait on the future until it completes.
    */
-  void wait() {
+  void wait(std::chrono::milliseconds timeout = kNoTimeout) {
```

Comment:
> This PR turned out to be more intense than I had anticipated, so for now I only managed to look at the common code and at the Gloo process group. I'll review the NCCL and the MPI ones later but perhaps you can start taking a look at the first comments?
> 
> One thing is that this PR could have been split into smaller ones. Reviewing an incremental stack is much easier than reviewing a single large PR. For example, I think the changes to Future, to add a timeout, could have been a standalone preliminary PR.

sounds good.

> A general design remark I have is that I think Work is now absolving two separate roles, and this is creating some confusion in the code: Work is used both by the producer (i.e., by the thread inside the process group that carries out the operation and sets the result of the future) and by the consumer (because it's returned to the user). It could make sense to split these two functions into two separate classes. This might also help with getting finally rid of Work, because the consumer-side class would need to be removed, but the producer-side one must stick around.

My thinking that Work and its subclasses is a producer class while Future would eventually become the only consumer class, do you propose a Future wrapper class to handle additional functionality, e.g. srcRank?